src/HAL/Hal.c

```diff
@@ -0,0 +1,314 @@
+            Nvic_setInterruptPriority(Nvic_Irq_Uart3, 5);
+            break;
+        case Uart_Id_4:
+            Nvic_enableInterrupt(Nvic_Irq_Uart4);
+            Nvic_setInterruptPriority(Nvic_Irq_Uart4, 5);
+            break;
+    }
+}
+
+void
+Hal_console_usart_init(void)
+{
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* CCFG_SYSIO = (uint32_t*)0x40088114;
+
+    // Assign the PB4 pin to the PIO controller (TDI is the default function)
+    *CCFG_SYSIO |= 1u << 4u;
+
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_PioB);
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_Usart1);
+
+    Pio pioB;
+    Pio_init(Pio_Port_B, &pioB);
+
+    Pio_Pin_Config pinConf;
+    pinConf.control = Pio_Control_PeripheralD;
+    pinConf.direction = Pio_Direction_Input;
+    pinConf.pull = Pio_Pull_None;
+    pinConf.filter = Pio_Filter_None;
+    pinConf.isMultiDriveEnabled = false;
+    pinConf.isSchmittTriggerDisabled = false;
+    pinConf.irq = Pio_Irq_None;
+
+    Pio_setPinsConfig(&pioB, PIO_PIN_4, &pinConf);
+
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_CR = (uint32_t*)0x40028000;
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_MR = (uint32_t*)0x40028004;
+    // cppcheck-suppress misra2012_11_4
+    uint32_t* US_BRGR = (uint32_t*)0x40028020;
+
+    *US_MR = 0;         // Clear
+    *US_MR |= 0u << 0;  // USART_MODE NORMAL
+    *US_MR |= 0u << 4;  // USCLKS PCK
+    *US_MR |= 3u << 6;  // CHRL 8BIT
+    *US_MR |= 0u << 8;  // SYNC Asynchronous
+    *US_MR |= 4u << 9;  // PAR No parity
+    *US_MR |= 0u << 12; // NBSTOP 1 stop bit
+    *US_MR |= 0u << 14; // CHMODE Normal mode
+    *US_MR |= 0u << 16; // MSBF MSB
+    *US_MR |= 0u << 17; // MODE9 CHRL defines length
+    *US_MR |= 0u << 18; // CLKO does not drive SCK pin
+    *US_MR |= 0u << 19; // OVER 16x oversampling
+    *US_MR |= 1u << 20; // INACK NSCK is not generated
+    *US_MR |= 0u << 21; // DSNACK don't care - no NACK
+    *US_MR |= 0u << 22; // VAR_SYNC MODSYNC defines sync
+    *US_MR |= 0u << 23; // INVDATA do not invert data
+    *US_MR |= 0u << 24; // MAX_ITERATION don't care - valid in ISO7816 protocol
+    *US_MR |= 0u << 28; // FILTER do not filter incomming data
+    *US_MR |= 0u << 29; // MAN manchester coding disabled
+    *US_MR |= 0u << 30; // MODSYNC don't care - manchester disabled
+    *US_MR |= 1u << 31; // ONEBIT 1 bit start frame delimiterx
+
+    // BaudRate = CLK / ((coarseDiv + fineDiv / 8) * 16)
+    uint32_t coarseDiv = SystemConfig_DefaultPeriphClock / (16u * USART_BAUD_RATE);
+    uint64_t fineDiv = 8uLL
+                       * (((uint64_t)SystemConfig_DefaultPeriphClock * 1000uLL / (16uLL * USART_BAUD_RATE))
+                          - (uint64_t)coarseDiv * 1000uLL);
+    fineDiv /= 1000uLL;
+
+    *US_BRGR = coarseDiv | ((uint32_t)fineDiv << 16u);
+    *US_CR = 1u << 6u; // Enable the transmitter
+}
+
+inline void
+waitForTransmitterReady(void)
+{
+    volatile uint32_t* const US_CSR = (uint32_t*)0x40028014;
```

Comment:
It would be nice if these numbers were named -> can we do a define for them? Possibly in something like Usart_Registers.h?